src/protobuf-net.Reflection/CSharpCodeGenerator.cs

```diff
@@ -839,23 +868,76 @@ protected string MakeRelativeName(GeneratorContext ctx, string typeName)
         private string GetTypeName(GeneratorContext ctx, string typeName)
         {
             string dataFormat = "";
+            NullabilityType? nullabilityType = null;
+            CompatibilityLevel? compatibilityLevel = null;
             bool isMap = false;
-            return GetTypeName(ctx, null, typeName, ref dataFormat, ref isMap, false);
+            return GetTypeName(ctx, null, typeName, ref dataFormat, ref nullabilityType, ref compatibilityLevel, ref isMap, false);
         }
 
-        private string GetTypeName(GeneratorContext ctx, FieldDescriptorProto field, string typeName, ref string dataFormat, ref bool isMap,
+        private string GetTypeName(
+            GeneratorContext ctx, 
+            FieldDescriptorProto field, 
+            string typeName, 
+            ref string dataFormat,
+            ref NullabilityType? nullabilityType,
+            ref CompatibilityLevel? compatibilityLevel,
+            ref bool isMap,
             bool nonNullable = false)
         {
+            if (ctx.EmitNullWrappers)
+            {
+                switch (typeName)
+                {
+                    case WellKnownTypeDouble:
+                        nullabilityType = NullabilityType.NullWrappedValue;
+                        return nonNullable ? "double" : "double?";
```

Comment:
in what scenario would be emit `double` (not `double?`) for a `DoubleValue` ? do we have tests for that scenario?

Reply:
that can never happen, agree. Returning `double?` no matter if the code flow passes `nonNullable` as `true` of `false`